### ISSUE TITLE
Non-RGB SPAN models

### ISSUE THREAD

**Issue by RunDevelopment** [muslll/neosr#25]
Hi @muslll!

I just read through the SPAN code again, and wondered whether span even supports anything other than RGB images as input. If I understand PyTorch tensors correctly, then this line:

https://github.com/muslll/neosr/blob/6973906c47bb62c2da00540e4dfce9a6ce10a49a/neosr/archs/span_arch.py#L239

will fail for non-RGB inputs because `self.mean` is defined like this:

https://github.com/muslll/neosr/blob/6973906c47bb62c2da00540e4dfce9a6ce10a49a/neosr/archs/span_arch.py#L222

and will always have 3 channels. 

So the `torch.Tensor(rgb_mean).view(1, 3, 1, 1)` should probably be changed to `torch.Tensor(rgb_mean).view(1, in_channels, 1, 1) `. Alternatively, we could also use the same approach as SwinIR:

https://github.com/muslll/neosr/blob/master/neosr/archs/swinir_arch.py#L775-L779

```py
        if in_chans == 3:
            rgb_mean = (0.4488, 0.4371, 0.4040)
            self.mean = torch.Tensor(rgb_mean).view(1, 3, 1, 1)
        else:
            self.mean = torch.zeros(1, 1, 1, 1)
```

Correction: The above suggested are not backwards compatible because of single-channel images are broadcasted. So we need to keep the current behavior for `in_chans in (1, 3)`.

What do you think?

**Comment by umzi2** [comment]
It at least runs the training and gives the expected result, today I can run the training of a 1 channel model and show the result.

**Comment by muslll** [comment]
Hi @RunDevelopment, this has been discussed a lot these few days in the EE discord. Consensus was that the `x = (x - self.mean) * self.img_range` forward affects stability. I thought about creating a bool for disabling it, which would also solve the grayscale training issue you mentioned. What do you think?

```
    def __init__(self,
                 num_in_ch=3,
                 num_out_ch=3,
                 feature_channels=48,
                 upscale=upscale,
                 bias=True,
                 norm=False,
                 img_range=1.0,
                 rgb_mean=(0.4488, 0.4371, 0.4040)
                 ):
        super(span, self).__init__()

        in_channels = num_in_ch
        out_channels = num_out_ch
        self.img_range = img_range
        self.mean = torch.Tensor(rgb_mean).view(1, 3, 1, 1)
        self.norm = norm

        self.conv_x = [...]
        
    def forward(self, x):
        if self.norm:
            self.mean = self.mean.type_as(x)
            x = (x - self.mean) * self.img_range

        [...]
        ```

**Comment by RunDevelopment** [comment]
Sounds good. I don't know much about AI, so I trust you and the others on EE when you say that removing this won't cause problems.

As for spandrel parameter detection: since this parameter is a boolean, this makes detection (in a backwards compatible way) easy. We can use the same trick I used [Real-CUGAN pro models](https://github.com/chaiNNer-org/spandrel/blob/689385114f377bcb29116e80b76b8e570856d772/src/spandrel/architectures/RealCUGAN/arch/upcunet_v3.py#L290-L294). Basically, we optionally register a small tensor as a buffer and the presence of the tensor determines the parameter value. 
In this case, we register the tensor only if `norm=False`. So the tensor will only be present on SPAN models without normalization, so everything is backwards compatible.

What do you think?

**Comment by muslll** [comment]
Looks good. [Commited](https://github.com/muslll/neosr/commit/8b87b7d564c61d65d82792c89e0fdccc2340003c) :+1:

**Comment by RunDevelopment** [comment]
Sorry for not making this clear enough @muslll, but the Real-CUGAN trick I mentioned has to be implemented by neosr as well. If neosr doesn't register the tensor to signify the value of `norm`, there is no way for us to detect this.

**Comment by RunDevelopment** [comment]
On that note: your CUGAN implementation for pro models has the same issue. Since the `pro` parameter isn't stored in the .pth, there is no way to detect them as pro models. They would even be [incompatible with the official Real-CUGAN code](https://github.com/bilibili/ailab/blob/2799af78ef105b414cc4b796c67c8511acdcdf6f/Real-CUGAN/upcunet_v3.py#L1263).

**Comment by muslll** [comment]
> Should I make a separate issue for this?

No need, I just commited [here](https://github.com/muslll/neosr/commit/972289d1db8e9e7452bee5b6c79e939712e1178c). I also changed the tensor to `isnorm` instead, to avoid conflict with the main function param.

**Comment by RunDevelopment** [comment]
> edit: [fix](https://github.com/muslll/neosr/commit/92c924d242c204021b85a560ea3e5eea7ca8a632)

Still wrong. As I said here:

> we register the tensor only if `norm=False`.

Right now, you are registering the tensor for the old behavior. But we need to register it for the new behavior and for it only.

**Comment by muslll** [comment]
I see. [Is this right now?](https://github.com/muslll/neosr/commit/90ead2c7b7300c95d2fbd28de0b4cd53dabb0e27).

**Comment by RunDevelopment** [comment]
Yes, this works! Thank you @muslll!

Also, you don't have to make a property and stuff like I did with Real-CUGAN. I just did it there because both the hyper parameter and buffer had the same name...

**Comment by muslll** [comment]
Great. Thanks :+1:
